### PR TITLE
Add a test build on NixOS to GitLab CI.

### PR DESCRIPTION
**Pull request by Zimmi48**
Unfortunately, we can't run the test-suite until #5975 is fixed.
We pin `default.nix` again to make the CI build predictable.
As in Windows builds, we need to override the default `before_script`. Shouldn't we get rid of this default and put it in the template instead?

**Kind:** infrastructure.

### CODE REVIEW

**Review by ejgallego — COMMENTED** [review]
I fail to see why this is useful. The purpose of Coq's CI is not to test every know distribution, much less a minoritary one.

I guess this should rather be on Nix's CI system.

**Comment by Zimmi48** [comment]
This was requested to me after I fixed yet another incompatibility that had been introduced in Coq.

**Comment by Zimmi48** [comment]
This test can also be very useful to prevent any "accidental" introduction of a new dependency.

Of course, those two advantages would be much stronger if we could run the test-suite too.

**Comment by ejgallego** [comment]
What part of Coq is this testing that is not tested elsewhere?



> This was requested to me after I fixed yet another incompatibility that had been introduced in Coq.

Can you say more?

> This test can also be very useful to prevent any "accidental" introduction of a new dependency.

Umm, why our current system does not prevent this?

File: .gitlab-ci.yml
Comment (ejgallego) [inline .gitlab-ci.yml]: This seems like a non-reproducible recipe.

File: default.nix
Comment (ejgallego) [inline default.nix]: Why this change? Is that fetching a concrete Nix image?

File: .gitlab-ci.yml
Comment (ejgallego) [inline .gitlab-ci.yml]: name this `distro:nixos` please.

**Comment by Zimmi48** [comment]
Sorry, I wasn't expecting this change to be controversial, so I hadn't really prepared arguments to defend it.

> The purpose of Coq's CI is not to test every know distribution, much less a minoritary one.

This is not any distribution, this is a test that everything works with the Nix package manager, which is of much more general interest. We already support this package manager by providing a `default.nix` so it's probably a good idea to test it as well. Using a NixOS image instead of an Ubuntu image with Nix on it allows for a much lighter image (download time is ridiculous) with zero unneeded package. That's why I'm saying that it can help to track Coq's dependencies better. For instance, when `git`, `rsync` and `which` were added in the test-suite. Not something that can be detected with a full-fledged Ubuntu image...

**Comment by Zimmi48** [comment]
It's funny because I was expecting instead the reaction to be: oh wow, how short and elegant it is to add a NixOS CI test.

**Comment by Zimmi48** [comment]
Here was the request to add this job by @SkySkimmer: https://github.com/coq/coq/pull/7484#issuecomment-388373819

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml]: We can install then run the test suite like gitlab does.

Comment (Zimmi48) [reply .gitlab-ci.yml]: That's not really a supported workflow in Nix. That would be a bit hackish. I prefer to wait and see if Dune is really going to solve this problem. If not, I'll propose to remove the problematic tests.

Comment (Zimmi48) [reply .gitlab-ci.yml]: Actually I was wrong: this is supported by Nix (https://nixos.org/nixpkgs/manual/#ssec-installCheck-phase). So I'll try that indeed.

Comment (SkySkimmer) [reply .gitlab-ci.yml]: > The installCheck phase checks whether the package was installed correctly by running its test suite against the installed directories. The default installCheck calls make installcheck. 

We could probably add such a target to the Makefile. It would simplify the gitlab yml too.

Comment (Zimmi48) [reply .gitlab-ci.yml]: Indeed. And this reinforces my conviction that it is wrong that currently the `check` target fails if configured with `-prefix`. Anyways, for now I'm remapping `installCheck` to use the `check` target and I'm getting pretty close to being able to run the test-suite successfully.

**Comment by Zimmi48** [comment]
In the nixpkgs manual, the `check` target is meant to be run after build but before install, and the `installCheck` target is meant to run after install. This is probably coming from a standard convention. We shouldn't provide a `check` target that fails if it is run before install. We could consider splitting our test-suite in two parts, one that is run by `check` and one that is run by `installCheck` (other solutions are possible).

**Comment by Zimmi48** [comment]
This is ready to merge if CI passes.

File: default.nix
Comment (SkySkimmer) [inline default.nix]: No -j?

**Comment by SkySkimmer** [comment]
Doesn't work: 
~~~
$ nix-build -f "$CI_PROJECT_URL/-/archive/$CI_COMMIT_SHA.tar.gz" -j2 -K
error: unrecognised flag '-f'
~~~

File: .gitlab-ci.yml
Comment (ejgallego) [reply .gitlab-ci.yml]: I know who has built this image but I don't see why is this relevant here. My question is still the same, this seems like a non-reproducible setup, doesn't it?

File: default.nix
Comment (SkySkimmer) [inline default.nix]: This includes `validate`, is that intended?
I don't think you can avoid setting the variables that the gitlab test suite job does to make it work with installed Coq, that's why I thought about having a separate installcheck target.

**Comment by ejgallego** [comment]
> Using a NixOS image instead of an Ubuntu image with Nix on it allows for a much lighter image (download time is ridiculous) with zero unneeded package. That's why I'm saying that it can help to track Coq's dependencies better. For instance, when git, rsync and which were added in the test-suite. Not something that can be detected with a full-fledged Ubuntu image...

Ubuntu base image is 80 MiB and contains just very basic packages, all in all I don't see an improvement in download time / image size. In fact, runners will cache the docker image but not the NixOS one so this part is not an improvement.

I made 3 questions above but I still got no answer, so tagging until they get answered.

**Review by ejgallego — CHANGES_REQUESTED** [review]
Answer my questions.

**Comment by ejgallego** [comment]
> For instance, when git, rsync and which were added in the test-suite. Not something that can be detected with a full-fledged Ubuntu image...

`git` and `rsync` are not part of the base Ubuntu image, is that the case for the NixOS image?

**Comment by ejgallego** [comment]
> It's funny because I was expecting instead the reaction to be: oh wow, how short and elegant it is to add a NixOS CI test.

Well my role as a reviewer is to ask very different questions, like "why do we need this?", "is this redudant with the test we already have?", "is the addition of a new job justified?", etc...

**Comment by ejgallego** [comment]
Indeed NixOS is special enough that we may want to run a build job.

**Comment by aspiwack** [comment]
@ejgallego I must admit that I have some doubts that such a trivial PR is worse that much emotion.

Nevertheless, I can provide some answers, I think

- It would be absolutely possible to use a Ubuntu image, install Nix on it, then call the same `default.nix`. It would have the same effect, but it does feel more complex than using this docker image.
- The `latest` tag is, indeed, strictly speaking non-reproducible. But because a specific version of nixpkgs is pinned for the build, the only issue would be if a newer version of nix could no longer read the version of nixpkgs under consideration. I don't know whether this ever happens, but we're talking guaranteed reproducibility for years. I doubt that we need more than that (and as far as reproducibility is concerned, since mirrors can disappear from the internet, it's probably very much theoretical to ask for more than a few years). The rest of the build will be identical, because it's the version of nixpkgs, not of nix, which drives the build. (this is the same reason why running in an Ubuntu image would make no difference)

File: .gitlab-ci.yml
Comment (Zimmi48) [reply .gitlab-ci.yml]: I'm unsure what exactly you are asking? Is this because of `latest`? What can I do to make it reproducible? Use a hash instead?

File: default.nix
Comment (Zimmi48) [reply default.nix]: That's a solution indeed, if you are not worried about the build time.

File: default.nix
Comment (Zimmi48) [reply default.nix]: Yes, it does include `validate`. I could restrict to `test-suite` if deemed sufficient.

File: .gitlab-ci.yml
Comment (ejgallego) [reply .gitlab-ci.yml]: From my part I can live with it if it is hard to improve, even if there are some components that worry such as openssl or git. WDYT @SkySkimmer ?

File: .gitlab-ci.yml
Comment (ejgallego) [reply .gitlab-ci.yml]: So maybe `pkg:nixos` ?

Based on what you say we would for example also add a test to build an opam package, so we would have `pkg:opam` and `pkg:nixos`. How does that sound?

**Comment by Zimmi48** [comment]
> git and rsync are not part of the base Ubuntu image, is that the case for the NixOS image?

I would expect they are either part of the base image, or that they are installed at some point (I haven't checked). Otherwise, how does GitLab pulls the repository?

I thought I had answered your questions but I'm going to try again:

> What part of Coq is this testing that is not tested elsewhere?

Similarly to when we test the compilation and test-suite with several versions of OCaml, this is not testing a different part but rather a different setup. BTW, it would be really easy to test different versions of OCaml with the Nix setup.

> Can you say more?

There is a number of incompatibilities that were fixed over time by either myself or @vbgl. Do you want the list?

> Umm, why our current system does not prevent this?

The `git` case is certainly one that was missed by our current system.

**Comment by ejgallego** [comment]
Thanks for the answers:

> > git and rsync are not part of the base Ubuntu image, is that the case for the NixOS image?
> 
> I would expect they are either part of the base image, or that they are installed at some point (I haven't checked). Otherwise, how does GitLab pulls the repository?

They are not part of the base image [check with `docker run -it ubuntu:bionic /bin/bash`, I think that gitlab pulls the repos from the runner, but I am not sure.

> > What part of Coq is this testing that is not tested elsewhere?
> 
> Similarly to when we test the compilation and test-suite with several versions of OCaml, this is not testing a different part but rather a different setup.

I see, more below.

> BTW, it would be really easy to test different versions of OCaml with the Nix setup.

Well in the current setup it is fairly easy too and in fact it has outperformed Nix in the past as OPAM has official status wrt OCaml. IMHO these kind of statements are a bit "misleading", as they imply NixOS brings and improvement here, but in fact it does not.

> > Can you say more?
> 
> There is a number of incompatibilities that were fixed over time by either myself or @vbgl. Do you want the list?

Indeed sorry as this is really my only question [remarks aside], so far I can see the `/bin/bash` case, what more differences in setup does NixOS help with? I don't need an exhaustive list, just an informal list of things that differ to get an idea.

> > Umm, why our current system does not prevent this?
> 
> The git case is certainly one that was missed by our current system.

`git` may be indeed special, but so far here is another part that where I fail to see an improvement wrt current setup. I could be wrong of course.

Summarizing, and until further correction, so far it seems to me that the only advantage NixOS does bring to our CI setup is to test against hardcoded paths.

That doesn't seem a lot, however, on the other hand, if we see NixOS as a packaging job indeed it does make sense to test the build script `default.nix` as it would make sense to test the opam files once they are auto generated, so it seems OK to have this job in the CI.

**Comment by SkySkimmer** [comment]
Probably git is pulled in by opam.

**Comment by SkySkimmer** [comment]
https://gitlab.com/SkySkimmer/coq/-/jobs/66654642
>Selecting previously unselected package git.
Preparing to unpack .../226-git_1%3a2.17.0-1ubuntu1_amd64.deb ...
Unpacking git (1:2.17.0-1ubuntu1) ...

**Comment by ejgallego** [comment]
> Probably git is pulled in by opam.

It should not, but indeed we have a bug in our Dockerfile, we should run apt with with `--no-install-recommends` as to really test strict dependencies.

**Comment by gares** [comment]
> And this reinforces my conviction that it is wrong that currently the |check| target fails if configured with |-prefix|.

@SkySkimmer please have a look at #7977

**Review by ejgallego — APPROVED** [review]
I would like to learn more about what is NixOS testing that the standard CI run does not do; apart from path hardcoding which is IMO useful.

Anyways, if we see this as a packaging job then indeed it does make sense to test our `default.nix` job, so green light here.

Note that we should qualify the job as to be `pkg:nixos` or something like that for when we introduce `pkg:opam`.

**Comment by Zimmi48** [comment]
FWIW pulling the Docker image takes less than 20 seconds (I've observed the build in real time - and GitLab helps by showing the time elapsed since the beginning).

> It should not, but indeed we have a bug in our Dockerfile, we should run apt with with --no-install-recommends as to really test strict dependencies.

Well, if you don't install `git` you have to change the "git strategy" as I did in the nixos job.

> Well in the current setup it is fairly easy too and in fact it has outperformed Nix in the past as OPAM has official status wrt OCaml. IMHO these kind of statements are a bit "misleading", as they imply NixOS brings and improvement here, but in fact it does not.

Don't worry I'm planning to show that Nix can outperform OPAM soon XD

> Indeed sorry as this is really my only question [remarks aside], so far I can see the /bin/bash case, what more differences in setup does NixOS help with? I don't need an exhaustive list, just an informal list of things that differ to get an idea.

So far, mostly this and the dependency on Python 2 that had gone unnoticed.

> Anyways, if we see this as a packaging job then indeed it does make sense to test our default.nix job, so green light here.

This is indeed how I see it.

> Note that we should qualify the job as to be pkg:nixos or something like that for when we introduce pkg:opam.

Agreed. I'll put `pkg:nix`. I prefer this to `distro:nixos`.

**Comment by ejgallego** [comment]
> Agreed. I'll put pkg:nix. I prefer this to distro:nixos.

Sounds good, we'll add opam and likely `deb` at some point.

**Comment by ejgallego** [comment]
Note that still there are some naming discrepancies [in the commit msg] wrt to Nix vs NixOS

**Comment by Zimmi48** [comment]
This is blocked by a consistent failure of the timing test. Here's the complete log: [timing.log](https://github.com/coq/coq/files/2156404/timing.log). The most interesting part being:

```
--- time-of-build-before.log.desired.processed	2018-07-02 18:29:54.820287819 +0000
+++ time-of-build-before.log.processed	2018-07-02 18:29:54.818287815 +0000
@@ -1,5 +1,11 @@
 COQDEP VFILES
 COQC Slowv
-Slow (real: , user: , sys: , mem: ko)
+------
+real	ms
+user	ms
+sys	ms
 COQC Fastv
-Fast (real: , user: , sys: , mem: ko)
+------
+real	ms
+user	ms
+sys	ms
```

It is working inside `nix-shell` though. I've tried adding `time` as an explicit dependency but this didn't change anything. @vbgl Could it be that the time command doesn't work in Nix builds?

In the meantime, I'm trying to understand why I don't manage to upload artifacts for failed test logs: https://gitlab.com/coq/coq/-/jobs/78880289
@SkySkimmer Any advice?

**Comment by SkySkimmer** [comment]
https://docs.gitlab.com/ee/ci/yaml/#artifacts-paths
>You can only use paths that are within the project workspace.

So /tmp/ stuff won't work. Maybe if you do `ln -s tmp /tmp/` you can use a relative `tmp/...` path? Try and see.

**Comment by Zimmi48** [comment]
Thanks! The error message wasn't very explicit ;-)
Anyways, instead of the hack you suggested, I've just told Nix to use the current directory as the `tmp` directory.

**Comment by Zimmi48** [comment]
The artifact uploading works now. So the only two blocking things are to figure out what is going on with the timing test, and to decide whether we keep things as they are now or go back to doing the check before install thanks to #7977.

**Comment by Zimmi48** [comment]
So in fact, the explanation is not that `time` doesn't work on Nix builds but that it outputs a different format when called as `time` or or as `/path/to/time`. Why? Because when calling `time`, it is a shell function that we are calling (https://askubuntu.com/a/434294/769627).

**Comment by gares** [comment]
Yep, time is a bash shell builtin, if we rely on the `gnu time` utility we should call it via `/usr/bin/time`.

**Comment by Zimmi48** [comment]
Alright this PR should be really ready now. I have fixed the timing tools on NixOS.
I have tried taking advantage of #7977 but I haven't managed to, so I'm going to keep things as they are now.

FTR the errors I got were:
```
Error: Cannot find a physical path bound to logical path matching suffix <> and prefix Coq.
```

@vbgl Could you review the changes to `default.nix`?

**Comment by gares** [comment]
> I have tried taking advantage of #7977 but I haven't managed to, so I'm going to keep things as they are now.

Can you tell more? What gave this error?

**Comment by SkySkimmer** [comment]
>FTR the errors I got were:

Without knowing what you did to adapt it's not really possible to investigate.

**Review by SkySkimmer — CHANGES_REQUESTED** [review]

**Comment by Zimmi48** [comment]
> Can you tell more? What gave this error?

I added this line:

```
checkFlags = [ "CAML_LD_LIBRARY_PATH=$(PWD)/kernel/byterun" "COQLIB=$(PWD)" ];
```

But looking at some other parts of the log, I can see:

```
Fatal error: exception Sys_error("/build/coq/test-suite/coq-makefile/arg/_test/plugins: No such file or directory")
```
I might have made wrong assumptions on the value of `$PWD` at check time. I don't know what other variable I could have used. Maybe @vbgl knows.

**Comment by SkySkimmer** [comment]
If I understand correctly that makes it run `make var=foo check` but then `var` is not exported, it's only set in the makefile. @gares was using `var=foo make check`.

**Comment by Zimmi48** [comment]
Ah, I didn't know the difference. I'm going to try again.

**Comment by Zimmi48** [comment]
In this branch: https://github.com/Zimmi48/coq/tree/save/ci/gitlab/nixos-2
I still get this error:

```
Makefile:423: warning: undefined variable 'TGTS'
Makefile:655: warning: undefined variable 'src/test_plugin_MLPACK_DEPENDENCIES'
Fatal error: cannot load shared library dllnums
Reason: dllnums.so: cannot open shared object file: No such file or directory
make[5]: *** [Makefile:668: theories/test.vo] Error 2
make[4]: *** [Makefile:329: all] Error 2
make[4]: Leaving directory '/build/coq/test-suite/coq-makefile/findlib-package/_test'
make[3]: *** [Makefile:752: byte] Error 2
make[3]: Leaving directory '/build/coq/test-suite/coq-makefile/findlib-package/_test'
==========> FAILURE <==========
    coq-makefile/findlib-package/run.sh...Error!
```

**Comment by gares** [comment]
This is what `CAML_LD_LIBRARY_PATH=$(PWD)/kernel/byterun` is supposed to fix

**Comment by gares** [comment]
Oh no sorry, it is not the vm, it is the .so part of bignum

**Comment by Zimmi48** [comment]
Are you sure it's not the num library from OCaml? In fact, I'm guessing that Nix had already set `CAML_LD_LIBRARY_PATH` and I overrode it. I should have appended to it.

**Comment by gares** [comment]
Yes, you should append the path leading to byterun

**Comment by gares** [comment]
Anyway, I think the nix way is a bit stretching it in this case.

The manpage of `ocamlrun` says that `CAML_LD_LIBRARY_PATH` is for *additional* paths to be scanned for .so objects...  and this is in line with that [LD_LIBRARY_PATH is for](http://tldp.org/HOWTO/Program-Library-HOWTO/shared-libraries.html): hacks and hijacking, not regular function.

I understand why they did it this way, but they should have asked the upstreams for for another variable name to avoid backstabbing their users: setting `(CAML_)LD_LIBRARY_PATH` should not hide stuff that is installed system wide, IMHO. [troll] What is next? Setting LD_PRELOAD system wide?

**Review by gares — CHANGES_REQUESTED** [review]
The `which time` part should be refined as follows:
```
gares@gargamel:~$ export PATH=/bin
gares@gargamel:~$ which time
gares@gargamel:~$ echo $?
1
gares@gargamel:~$ which time || echo time
time
gares@gargamel:~$ export PATH=/bin:/usr/bin/
gares@gargamel:~$ which time || echo time
/usr/bin//time
```

File: default.nix
Comment (vbgl) [reply default.nix]: I guess we discussed that already; and we were in favor of not pinning within `default.nix` (which is used daily by developers). If you really need pinning, please do so in a separate file that has to be explicitly called; in the CI script, you may call `nix-build -f reproducible.nix …`, but when working as usual, I’ll only run plain `nix-shell`.

**Comment by Zimmi48** [comment]
I'm almost back to where I was with `installCheck`, only the timing test is still failing, this time with:

```
Makefile:71: warning: undefined variable '*'

--- time-of-build-before.log.desired.processed	2018-07-03 14:07:28.134304031 +0000
+++ time-of-build-before.log.processed	2018-07-03 14:07:28.133304032 +0000
@@ -1,5 +1,2 @@
-COQDEP VFILES
-COQC Slowv
 Slow (real: , user: , sys: , mem: ko)
-COQC Fastv
 Fast (real: , user: , sys: , mem: ko)
==========> FAILURE <==========
    coq-makefile/timing/run.sh...Error!
```

**Comment by Zimmi48** [comment]
@gares I don't think you have reviewed the last version of the PR. There is no `which time` anymore. I'm using `command time` as was requested by @SkySkimmer.

**Comment by aspiwack** [comment]
Is testing the test-suite that useful in this particular job?

**Comment by SkySkimmer** [comment]
Is that undefined variable expected?

File: default.nix
Comment (vbgl) [inline default.nix]: Why do we need `python2`? Can we contemplate porting that to python 3?

(I’m looking at the `python3` a few lines below…)

**Comment by Zimmi48** [comment]
> Is that undefined variable expected?

I would guess not.

**Review by gares — APPROVED** [review]

**Review by vbgl — CHANGES_REQUESTED** [review]

**Comment by Zimmi48** [comment]
> Is testing the test-suite that useful in this particular job?

Yes, I wouldn't have discovered that the timing tools were broken on NixOS if I hadn't try to have the test-suite pass. Anyways, we are nearly there now and we have two options: find out what's going on with the above mentioned error of the timing test and this undefined variable, or merge the current version which is using the `installCheck` phase and which passed CI.

**Comment by aspiwack** [comment]
> we have two options: find out what's going on with the above mentioned error of the timing test and this undefined variable, or merge the current version which is using the installCheck phase and which passed CI.

Merge first, improve later. I see no reason of delaying this.

**Comment by Zimmi48** [comment]
> I guess we discussed that already; and we were in favor of not pinning within default.nix (which is used daily by developers). If you really need pinning, please do so in a separate file that has to be explicitly called; in the CI script, you may call nix-build -f reproducible.nix …, but when working as usual, I’ll only run plain nix-shell.

That's not exactly how I reminded the conclusion. In my mind, the conclusion was: we are not going to keep pinning now for something that will happen in an hypothetical future. And the hypothetical future is now.
I don't understand why we would not like to pin by default given that `default.nix` provides the option to override this default. Here are my reasons:

1. We are testing the package that we are providing (and that people can install by doing `nix-env -if https://github.com/coq/coq/archive/master.zip`).
2. If we are not pinning the version that we are using, we'll forget to update it.
3. What's the problem with using a pinned version for `nix-shell` anyways?
4. And if you really don't want to use the pin version and that typing `nix-shell --arg pkgs "import <nixpkgs> {}"` is too hard, you can set up a bash alias `coq-dev: nix-shell --arg pkgs "import <nixpkgs> {}`. FTR my personal default nixpkgs channel is not nixpkgs-unstable and is too old for Coq development so I have to do something like this already and it's not a big deal.

> Merge first, improve later. I see no reason of delaying this.

I usually agree but this PR is changing `doCheck` into `doInstallCheck`. If the plan is to go back to `doCheck` immediately after, I prefer to do it at once. That being said, I have no idea how to fix the last issue I reported, so I'm proposing to merge as-is unless someone has an idea.

**Comment by maximedenes** [comment]
There are many things I don't follow here, and I don't have time answer immediately, please don't pin again by default meanwhile.

**Comment by vbgl** [comment]
Can we pin in default.nix and un-pin in shell.nix? So that there are good defaults both for developers and for end-users?

**Comment by maximedenes** [comment]
> Can we pin in default.nix and un-pin in shell.nix? So that there are good defaults both for developers and for end-users?

That would work for me.

**Comment by Zimmi48** [comment]
> Can we pin in default.nix and un-pin in shell.nix? So that there are good defaults both for developers and for end-users?

Yes and that would be fine to me as a way to make you and @maximedenes happy. So I'm going to push a commit doing this even if I still fail to see why unpinning does to help developers.

**Comment by vbgl** [comment]
I think that good defaults for developers **is** important.

That being said, I’m not completely convinced that what I just suggested completely solves the issue of pinning. Can you please clarify how/when/why/… will the reference to `nixpkgs` be updated?

**Comment by Zimmi48** [comment]
> I think that good defaults for developers is important.

I agree. I just disagree on what these good defaults are.

> That  being said, I’m not completely convinced that what I just suggested  completely solves the issue of pinning. Can you please clarify  how/when/why/… will the reference to nixpkgs be updated?

Don't worry, I'll personally use the pinned version from `default.nix` instead of the unpinned version from `shell.nix` so I'll update the pin as often as I update my nixpkgs-unstable channel (every other month or so).

**Comment by vbgl** [comment]
Of course, if the pinned version always matches your default version, you won’t understand the issue…

**Comment by Zimmi48** [comment]
So @SkySkimmer @vbgl are we gonna merge this?

**Comment by SkySkimmer** [comment]
Why is it building camlp5, is there not a prebuilt version?

File: default.nix
Comment (SkySkimmer) [inline default.nix]: Is building byte simultaneously OK?

**Review by SkySkimmer — APPROVED** [review]
The log is very bloated by nix complaining about "wrong elf type".

**Comment by Zimmi48** [comment]
> Why is it building camlp5, is there not a prebuilt version?

I guess because we are not using the default OCaml version from nixpkgs. Anyways my next PR will add caching.

**Comment by Zimmi48** [comment]
Only CI failure is mtac2, which is broken with current master.

**Comment by Zimmi48** [comment]
@vbgl I've addressed your request. Can you approve the PR?

**Review by vbgl — APPROVED** [review]

**Comment by vbgl** [comment]
Théo, you haven’t answered my questions: why do we pin the reference to nixpkgs? to which version? when do we update and how?

I nonetheless recommend merging.

**Comment by Zimmi48** [comment]
> Théo, you haven’t answered my questions: why do we pin the reference to nixpkgs? to which version? when do we update and how?

Oh! That's the second time in this PR that I think I have answered the questions and the reviewer says I haven't :scream: 
Why pin? For reproducibility. In fact, the NixOS image I use doesn't even contains a nixpkgs channel and encourages its users to pin (https://hub.docker.com/r/nixorg/nix/)

Which version? The latest version of nixpkgs when the pin is updated.

When update? As often as we want to. In the beginning, I'll update by hand. When I'll proceed to implement the proposal #6724, I'll include it in the regular (weekly) automatic update.

**Comment by Zimmi48** [comment]
@SkySkimmer Can you merge?

**Comment by Zimmi48** [comment]
FTR here is the diff to go back to check before install which almost works except for the problem with the timing test mentioned above:

```diff
diff --git a/default.nix b/default.nix
index 1be274081..c76526266 100644
--- a/default.nix
+++ b/default.nix
@@ -16,7 +16,7 @@
 # $ nix-build
 # at the root of the Coq repository.
 # nix-build also supports the --arg option, so you will be able to do:
-# $ nix-build --arg doInstallCheck false
+# $ nix-build --arg doCheck false
 # if you want to speed up things by not running the test-suite.
 # Once the build is finished, you will find, in the current directory,
 # a symlink to where Coq was installed.
@@ -27,9 +27,10 @@
       sha256 = "1lzvp3md0hf6kp2bvc6dbzh40navlyd51qlns9wbkz6lqk3lgf6j";
     }) {})
 , ocamlPackages ? pkgs.ocaml-ng.ocamlPackages_4_06
+, buildByte ? true
 , buildIde ? true
 , buildDoc ? true
-, doInstallCheck ? true
+, doCheck ? true
 }:
 
 with pkgs;
@@ -46,14 +47,15 @@ stdenv.mkDerivation rec {
   ++ (with ocamlPackages; [ ocaml findlib camlp5_strict num ])
   ++ optional buildIde ocamlPackages.lablgtk
   ++ optionals buildDoc [
+
     # Sphinx doc dependencies
     pkgconfig (python3.withPackages
       (ps: [ ps.sphinx ps.sphinx_rtd_theme ps.pexpect ps.beautifulsoup4
              ps.antlr4-python3-runtime ps.sphinxcontrib-bibtex ]))
     antlr4
   ]
-  ++ optionals doInstallCheck (
-    # Test-suite dependencies
+  ++ optionals doCheck ( # Test-suite dependencies
+
     # ncurses is required to build an OCaml REPL
     optional (!versionAtLeast ocaml.version "4.07") ncurses
     ++ [ ocamlPackages.ounit rsync which ]
@@ -72,18 +74,19 @@ stdenv.mkDerivation rec {
 
   prefixKey = "-prefix ";
 
-  buildFlags = [ "world" "byte" ] ++ optional buildDoc "doc-html";
+  buildFlags =
+   [ "world" ] ++ optional buildByte "byte" ++ optional buildDoc "doc-html";
 
   installTargets =
-    [ "install" "install-byte" ] ++ optional buildDoc "install-doc-html";
+    [ "install"] ++ optional buildByte "install-byte"
+    ++ optional buildDoc "install-doc-html";
 
-  inherit doInstallCheck;
+  inherit doCheck;
 
-  preInstallCheck = ''
+  preCheck = ''
     patchShebangs tools/
     patchShebangs test-suite/
+    export CAML_LD_LIBRARY_PATH=$PWD/kernel/byterun:$CAML_LD_LIBRARY_PATH
+    export COQLIB=$PWD
   '';
-
-  installCheckTarget = [ "check" ];
-
 }
```

**Pull request merged**
